### ISSUE TITLE
General Haxe conversion

### ISSUE THREAD

**Issue by piotrzarzycki21** [Moonshine-IDE/Moonshine-IDE#1035]
As part of this issue we will manage current work on converting business logic - mostly non related to UI or tightly related. All work is happening in branch [features/haxeconversion](https://github.com/Moonshine-IDE/Moonshine-IDE/tree/features/haxeconversion) and soon will be merged to master.

**Comment by Igazine** [comment]
Thank you

**Comment by Igazine** [comment]
FeathersUI has been updated to version 1.0.0-rc.1. The Ant build file pulls the latest one, so no action is needed on our end, just note that the conversion goes on using the latest version of FeathersUI

**Comment by Igazine** [comment]
Yes, sure, I'll put together a list of what has been converted already and what part of the software may be affected

**Comment by piotrzarzycki21** [comment]
@Igazine how is going with this list ?

**Comment by Igazine** [comment]
@piotrzarzycki21 Please check https://github.com/Moonshine-IDE/Moonshine-IDE/tree/features/haxeconversion/ide/MoonshineGUICore/docs/conversionstatus for changelog

And please let me know if you need anything else. Thank you

**Comment by piotrzarzycki21** [comment]
> I'm fine with a merge if we are confident that no feature conversion missed out, or at least we aware the part of the feature-set still needs a conversion.

I don't think I'm confident - until I start using daily. This is my point. We have to stop holding things on branches for months. If I hear that it is working, stable - I would like to see it on master and move on.

**Comment by Aszusz** [comment]
I started testing Haxe Conversion Branch Windows Build 4
All top-level menus (File, Edit, View) are inoperable. Here's the call stack:
```
 TypeError: Error #1034
: 	at actionScripts.ui.menu.vo::CustomMenuItem/get_submenu()
: 	at actionScripts.ui.menu.vo::CustomMenuItem/get submenu()
: 	at actionScripts.ui.menu.renderers::MenuRenderer/drawMenuState()
: 	at actionScripts.ui.menu.renderers::MenuRenderer/commitProperties()
: 	at mx.core::UIComponent/validateProperties()
: 	at mx.managers::LayoutManager/validateClient()
: 	at mx.managers::PopUpManagerImpl/addPopUp()
: 	at mx.managers::PopUpManager$/addPopUp()
: 	at actionScripts.ui.menu::MenuModel/positionMenu()
: 	at actionScripts.ui.menu::MenuModel/displayMenu()
: 	at actionScripts.ui.menu::MenuBar/displayMenuAt()
: 	at actionScripts.ui.menu::MenuBar/barItemOpenMenu()
: 	at flash.events::EventDispatcher/dispatchEvent()
: 	at mx.core::UIComponent/dispatchEvent()
: 	at mx.managers::SystemManager/mouseEventHandler()
```

**Comment by Igazine** [comment]
Somehow the menu builder receives an `Event` object instead of a `CustomMenuItem`, but only on Windows. I suspect it's related to how the menus are being built/assembled in the code.

It's fixed for now, but I need to look deeper into these irregularities.

@Aszusz I think after the build, it's ready to be tested. Thank you

**Comment by rat-moonshine** [comment]
@Igazine please give me some more detail to clear my understanding on the application menu subject. I was in impression that we discussed the application menus requires Haxe Extension to work. Or, did I missed something (?)

1. Is the application menus are working both on macOS and Windows?
2. Did you converted `MenuItem` for Windows where the menu is always attached to the application?
3. Do we still has requirement for Haxe Extension for the menus on macOS?

**Comment by Igazine** [comment]
@rat-moonshine This issue is unrelated to Native Menus, it's a coding issue that resides in the current ActionScript code.

1. Yes (not sure if I understand the question)
2. MenuItem is not converted yet, the actual application menu, context menu etc. system menus are going to need to be implemented as native extensions as we discussed earlier. Currently the menus are either ActionScript-based or Flash/AIR externs defined in Haxe code
3. Yes, we still need native extensions for every targeted platform (macOS, Windows, Linux). I've already started working on it, but in the meantime I'm looking for someone who can help us out with this task. This part will probably be one of the last tasks regarding the Moonshine Haxe conversion

**Comment by rat-moonshine** [comment]
Ah! Thank you for the clarification. Sorry, I still often forgot that Moonshine still running on AIR shell, and converting its AS3 codes only.

**Comment by Igazine** [comment]
Yes, up until to the point where Moonshine can be built as a native application, we're partially relying on Flash/AIR API. Even if we had a fully working cross-platform Native Menu implementation today, we still couldn't use it in AIR.

As the conversion is going forward, we're less and less dependent on Flash/AIR API. Josh has published some new Haxe libraries (e.g. AMFIO) that can be directly used in the code, so those were huge steps going forward.

The NativeMenu part is tricky because of the following reasons:
* There's no intention from the Lime/OpenFL team to make this API available in the frameworks, because their main goal is that whatever they implement, it should be flawlessly supported on all platforms (Native, iOS, Android, HTML, Hashlink, Neko etc), so anything that's platform-specific (like NativeMenus) gets way less attention
* Each targeted platforms handles native menus differently. While menus are part of the system APIs on macOS and Windows (hence it's relatively easy to implement), it's a whole different story on Linux. There are several kinds of desktop environments and each of them are using different SDKs/libraries to display window menus, context menus etc., so it'll be much tougher, but we'll discuss our options when we get there.

**Comment by rat-moonshine** [comment]
I had an intention to look into macOS menus myself for our GBAuth conversion as in a Haxe Extension. But I admit I still didn't had much time to look into this, and I still have more things to iron-out first. In any way, my impression is this (menu implementation as extension) going to be lengthy as this will involve understanding many things and learn.

**Comment by Aszusz** [comment]
Haxe Conversion Build incorrectly recognizes and cannot use Apache Royale SDKs. I tested this on Win and compared it to the latest stable which works correctly. I've seen this issue also on macOS.

Apache Royale SDKs are listed in the settings as (null only) instead of (js only):

![image](https://user-images.githubusercontent.com/7197014/191533609-bd39280b-8975-4acc-bc85-9c31d03d95b9.png)

When I try to build Royale Browser Project using this setup I get this error:
```
: This SDK does not contains playerglobal.swc in frameworks\libs\player\{version}\playerglobal.swc. Download playerglobal here
```
